File: stewardpro/stewardpro/report/financial_summary/financial_summary.js
// Copyright (c) 2024, StewardPro Team and contributors
// For license information, please see license.txt

frappe.query_reports["Financial Summary"] = {
	"filters": [
		{
			"fieldname": "period",
			"label": __("Period"),
			"fieldtype": "Select",
			"options": "Current Month\nQuarter\nYear to Date\nCustom",
			"default": "Current Month"
		}
	],
	
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		// Format headers
		if (column.fieldname == "category" && data && data.category) {
			if (data.category.includes("<b>")) {
				value = data.category;
			}
		}
		
		// Color code percentage changes
		if (column.fieldname == "month_change" || column.fieldname == "year_change") {
			if (data && typeof data[column.fieldname] === 'number') {
				const change = data[column.fieldname];
				if (change > 0) {
					value = `<span class="text-success">${value}</span>`;
				} else if (change < 0) {
					value = `<span class="text-danger">${value}</span>`;
				}
			}
		}
		
		return value;
	},
	
	"onload": function(report) {
		// Initialize chart utilities
		if (typeof stewardpro !== 'undefined' && stewardpro.charts) {
			stewardpro.charts.add_chart_selector(report, get_financial_chart_data, {
				default_chart_type: 'pie',
				height: 350
			});
		}

		// Add custom chart buttons
		report.page.add_inner_button(__("Income vs Expenses"), function() {
			show_income_expense_chart(report);
		}, __("Charts"));

		report.page.add_inner_button(__("Income Breakdown"), function() {
			show_income_breakdown_chart(report);
		}, __("Charts"));

		report.page.add_inner_button(__("Expense Analysis"), function() {
			show_expense_breakdown_chart(report);
		}, __("Charts"));

		report.page.add_inner_button(__("Trend Comparison"), function() {
			show_trend_comparison_chart(report);
		}, __("Charts"));

		report.page.add_inner_button(__("Financial Dashboard"), function() {
			show_financial_dashboard(report);
		}, __("Charts"));

		// Add export button
		report.page.add_inner_button(__("Export Summary"), function() {
			frappe.query_report.export_report();
		});
	},

	"after_datatable_render": function(report) {
		// Auto-refresh chart when data changes
		setTimeout(() => {
			if (report.data && report.data.length > 0 && typeof stewardpro !== 'undefined' && stewardpro.charts) {
				stewardpro.charts.refresh_chart(report, get_financial_chart_data);
			}
		}, 500);
	}
};

// Chart data processing functions
function get_financial_chart_data(data, filters, chart_type) {
	if (!data || !data.length) return null;

	chart_type = chart_type || 'pie';

	switch (chart_type) {
		case 'pie':
		case 'donut':
			return get_income_expense_pie_data(data);
		case 'line':
		case 'area':
			return get_financial_trend_data(data, true);
		case 'bar':
		default:
			return get_financial_trend_data(data, false);
	}
}

function get_income_expense_pie_data(data) {
	let total_income = 0;
	let total_expenses = 0;

	data.forEach(row => {
		const category = row.category || "";
		const current_month = row.current_month || 0;

		if (category === "Tithes" || category === "Regular Offerings" || category === "Special Offerings") {
			total_income += current_month;
		} else if (category.includes("Expenses")) {
			total_expenses += current_month;
		}
	});

	return {
		title: __('Income vs Expenses (Current Month)'),
		data: {
			labels: [__('Income'), __('Expenses')],
			datasets: [{
				values: [total_income, total_expenses]
			}]
		}
	};
}

function get_financial_trend_data(data, smooth_line = false) {
	const income_categories = [];
	const current_month_values = [];
	const previous_month_values = [];
	const ytd_values = [];
	
	data.forEach(row => {
		const category = row.category || "";
		if (category === "Tithes" || category === "Regular Offerings" || category === "Special Offerings") {
			income_categories.push(category);
			current_month_values.push(row.current_month || 0);
			previous_month_values.push(row.previous_month || 0);
			ytd_values.push(row.year_to_date || 0);
		}
	});
	
	return {
		title: __('Financial Trends Comparison'),
		data: {
			labels: income_categories,
			datasets: [
				{
					name: __('Current Month'),
					values: current_month_values
				},
				{
					name: __('Previous Month'),
					values: previous_month_values
				},
				{
					name: __('Year to Date'),
					values: ytd_values
				}
			]
		},
		options: {
			lineOptions: {
				hideDots: smooth_line ? 0 : 1,
				heatline: smooth_line ? 1 : 0
			}
		}
	};
}

// Chart view functions
function show_income_expense_chart(report) {
	if (!report.data || !report.data.length) {
		frappe.msgprint(__('No data available for chart'));
		return;
	}
	
	const chart_data = get_income_expense_pie_data(report.data);
	show_chart_modal(chart_data, 'pie');
}

function show_income_breakdown_chart(report) {
	if (!report.data || !report.data.length) {
		frappe.msgprint(__('No data available for chart'));
		return;
	}

	frappe.call({
		method: 'stewardpro.stewardpro.report.financial_summary.financial_summary.get_income_breakdown_chart',
		args: {
			data: JSON.stringify(report.data),
			filters: report.get_values()
		},
		callback: function(r) {
			if (r.message) {
				r.message.title = __('Income Breakdown');
				show_chart_modal(r.message, 'donut');
			} else {
				frappe.msgprint(__('No income data available to display'));
			}
		}
	});
}

function show_expense_breakdown_chart(report) {
	if (!report.data || !report.data.length) {
		frappe.msgprint(__('No data available for chart'));
		return;
	}

	frappe.call({
		method: 'stewardpro.stewardpro.report.financial_summary.financial_summary.get_expense_breakdown_chart',
		args: {
			data: JSON.stringify(report.data),
			filters: report.get_values()
		},
		callback: function(r) {
			if (r.message) {
				r.message.title = __('Expense Analysis');
				show_chart_modal(r.message, 'bar');
			} else {
				frappe.msgprint(__('No expense data available to display'));
			}
		}
	});
}

function show_trend_comparison_chart(report) {
	if (!report.data || !report.data.length) {
		frappe.msgprint(__('No data available for chart'));
		return;
	}

	frappe.call({
		method: 'stewardpro.stewardpro.report.financial_summary.financial_summary.get_trend_comparison_chart',
		args: {
			data: JSON.stringify(report.data),
			filters: report.get_values()
		},
		callback: function(r) {
			if (r.message) {
				r.message.title = __('Trend Comparison');
				show_chart_modal(r.message, 'bar');
			} else {
				frappe.msgprint(__('No trend data available to display'));
			}
		}
	});
}

function show_financial_dashboard(report) {
	if (!report.data || !report.data.length) {
		frappe.msgprint(__('No data available for dashboard'));
		return;
	}
	
	// Calculate financial metrics
	let total_income_current = 0;
	let total_income_previous = 0;
	let total_income_ytd = 0;
	let total_expenses_current = 0;
	let total_expenses_ytd = 0;
	
	report.data.forEach(row => {
		const category = row.category || "";
		const current = row.current_month || 0;
		const previous = row.previous_month || 0;
		const ytd = row.year_to_date || 0;

		if (category === "Tithes" || category === "Regular Offerings" || category === "Special Offerings") {
			total_income_current += current;
			total_income_previous += previous;
			total_income_ytd += ytd;
		} else if (category.includes("Expenses")) {
			total_expenses_current += current;
			total_expenses_ytd += ytd;
		}
	});
	
	const net_current = total_income_current - total_expenses_current;
	const net_ytd = total_income_ytd - total_expenses_ytd;
	const income_growth = total_income_previous > 0 ? 
		((total_income_current - total_income_previous) / total_income_previous * 100) : 0;
	
	// Create dashboard dialog
	const dialog = new frappe.ui.Dialog({
		title: __('Financial Dashboard'),
		size: 'extra-large',
		fields: [
			{
				fieldtype: 'HTML',
				fieldname: 'dashboard_html'
			}
		]
	});
	
	const dashboard_html = `
		<div class="row">
			<div class="col-md-3">
				<div class="card text-center">
					<div class="card-body">
						<h3 class="text-success">${frappe.format(total_income_current, {fieldtype: 'Currency'})}</h3>
						<p class="card-text">Current Month Income</p>
						<small class="${income_growth >= 0 ? 'text-success' : 'text-danger'}">
							${income_growth >= 0 ? '↑' : '↓'} ${Math.abs(income_growth).toFixed(1)}%
						</small>
					</div>
				</div>
			</div>
			<div class="col-md-3">
				<div class="card text-center">
					<div class="card-body">
						<h3 class="text-danger">${frappe.format(total_expenses_current, {fieldtype: 'Currency'})}</h3>
						<p class="card-text">Current Month Expenses</p>
					</div>
				</div>
			</div>
			<div class="col-md-3">
				<div class="card text-center">
					<div class="card-body">
						<h3 class="${net_current >= 0 ? 'text-success' : 'text-danger'}">${frappe.format(net_current, {fieldtype: 'Currency'})}</h3>
						<p class="card-text">Net Current Month</p>
					</div>
				</div>
			</div>
			<div class="col-md-3">
				<div class="card text-center">
					<div class="card-body">
						<h3 class="${net_ytd >= 0 ? 'text-success' : 'text-danger'}">${frappe.format(net_ytd, {fieldtype: 'Currency'})}</h3>
						<p class="card-text">Net Year to Date</p>
					</div>
				</div>
			</div>
		</div>
		<div class="row mt-4">
			<div class="col-md-6">
				<div class="card">
					<div class="card-header"><strong>Income Summary (YTD)</strong></div>
					<div class="card-body">
						<div class="progress mb-2">
							<div class="progress-bar bg-success" style="width: 100%"></div>
						</div>
						<h4 class="text-success">${frappe.format(total_income_ytd, {fieldtype: 'Currency'})}</h4>
					</div>
				</div>
			</div>
			<div class="col-md-6">
				<div class="card">
					<div class="card-header"><strong>Expenses Summary (YTD)</strong></div>
					<div class="card-body">
						<div class="progress mb-2">
							<div class="progress-bar bg-danger" style="width: ${total_income_ytd > 0 ? (total_expenses_ytd / total_income_ytd * 100) : 0}%"></div>
						</div>
						<h4 class="text-danger">${frappe.format(total_expenses_ytd, {fieldtype: 'Currency'})}</h4>
						<small>${total_income_ytd > 0 ? (total_expenses_ytd / total_income_ytd * 100).toFixed(1) : 0}% of income</small>
					</div>
				</div>
			</div>
		</div>
	`;
	
	dialog.fields_dict.dashboard_html.$wrapper.html(dashboard_html);
	dialog.show();
}

function show_chart_modal(chart_data, chart_type) {
	const dialog = new frappe.ui.Dialog({
		title: chart_data.title || __('Chart'),
		size: 'large',
		fields: [
			{
				fieldtype: 'HTML',
				fieldname: 'chart_html'
			}
		]
	});

	dialog.show();

	setTimeout(() => {
		const chart_config = {
			data: chart_data.data,
			type: chart_type,
			height: 300,
			colors: ['#4CAF50', '#2196F3', '#FF9800', '#9C27B0']
		};

		new frappe.Chart(dialog.fields_dict.chart_html.$wrapper[0], chart_config);
	}, 200);
}

File: stewardpro/stewardpro/report/financial_summary/financial_summary.py
# Copyright (c) 2024, StewardPro Team and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate
from frappe.query_builder.functions import Sum


def execute(filters=None):
	columns = get_columns()
	data = get_data(filters)
	chart = get_chart_data(data, filters)

	return columns, data, None, chart


def get_columns():
	return [
		{
			"label": _("Category"),
			"fieldname": "category",
			"fieldtype": "Data",
			"width": 200
		},
		{
			"label": _("Current Month"),
			"fieldname": "current_month",
			"fieldtype": "Currency",
			"width": 150
		},
		{
			"label": _("Previous Month"),
			"fieldname": "previous_month",
			"fieldtype": "Currency",
			"width": 150
		},
		{
			"label": _("Year to Date"),
			"fieldname": "year_to_date",
			"fieldtype": "Currency",
			"width": 150
		},
		{
			"label": _("Previous Year"),
			"fieldname": "previous_year",
			"fieldtype": "Currency",
			"width": 150
		},
		{
			"label": _("% Change (Month)"),
			"fieldname": "month_change",
			"fieldtype": "Percent",
			"width": 120
		},
		{
			"label": _("% Change (Year)"),
			"fieldname": "year_change",
			"fieldtype": "Percent",
			"width": 120
		}
	]


def get_data(filters):
	from frappe.utils import add_months, get_first_day, get_last_day
	
	# Date calculations
	today = getdate()
	current_month_start = get_first_day(today)
	current_month_end = get_last_day(today)
	
	previous_month_start = get_first_day(add_months(today, -1))
	previous_month_end = get_last_day(add_months(today, -1))
	
	year_start = getdate(f"{today.year}-01-01")
	previous_year_start = getdate(f"{today.year - 1}-01-01")
	previous_year_end = getdate(f"{today.year - 1}-12-31")
	
	data = []
	
	# Income Section
	data.append({
		"category": "<b>INCOME</b>",
		"current_month": "",
		"previous_month": "",
		"year_to_date": "",
		"previous_year": "",
		"month_change": "",
		"year_change": ""
	})
	
	# Tithes
	tithe_data = get_tithe_data(current_month_start, current_month_end, 
								previous_month_start, previous_month_end,
								year_start, today, previous_year_start, previous_year_end)
	data.append(tithe_data)
	
	# Offerings
	offering_data = get_offering_data(current_month_start, current_month_end,
									  previous_month_start, previous_month_end,
									  year_start, today, previous_year_start, previous_year_end)
	data.append(offering_data)
	
	# Special Offerings
	special_data = get_special_offerings_data(current_month_start, current_month_end,
											  previous_month_start, previous_month_end,
											  year_start, today, previous_year_start, previous_year_end)
	data.append(special_data)

	# Department Income
	dept_income_data = get_department_income_data(current_month_start, current_month_end,
												  previous_month_start, previous_month_end,
												  year_start, today, previous_year_start, previous_year_end)
	data.extend(dept_income_data)

	# Total Income
	total_income = calculate_total_income(data)
	data.append(total_income)
	
	# Expenses Section
	data.append({
		"category": "",
		"current_month": "",
		"previous_month": "",
		"year_to_date": "",
		"previous_year": "",
		"month_change": "",
		"year_change": ""
	})
	
	data.append({
		"category": "<b>EXPENSES</b>",
		"current_month": "",
		"previous_month": "",
		"year_to_date": "",
		"previous_year": "",
		"month_change": "",
		"year_change": ""
	})
	
	# Department Expenses
	expense_data = get_expense_data(current_month_start, current_month_end,
									previous_month_start, previous_month_end,
									year_start, today, previous_year_start, previous_year_end)
	data.extend(expense_data)

	return data


def get_tithe_data(cm_start, cm_end, pm_start, pm_end, yt_start, yt_end, py_start, py_end):
	tithes_table = frappe.qb.DocType("Tithes and Offerings")

	# Current month
	current_month = (
		frappe.qb.from_(tithes_table)
		.select(Sum(tithes_table.tithe_amount))
		.where(
			(tithes_table.date >= cm_start) &
			(tithes_table.date <= cm_end) &
			(tithes_table.docstatus == 1)
		)
	).run()[0][0] or 0

	# Previous month
	previous_month = (
		frappe.qb.from_(tithes_table)
		.select(Sum(tithes_table.tithe_amount))
		.where(
			(tithes_table.date >= pm_start) &
			(tithes_table.date <= pm_end) &
			(tithes_table.docstatus == 1)
		)
	).run()[0][0] or 0

	# Year to date
	year_to_date = (
		frappe.qb.from_(tithes_table)
		.select(Sum(tithes_table.tithe_amount))
		.where(
			(tithes_table.date >= yt_start) &
			(tithes_table.date <= yt_end) &
			(tithes_table.docstatus == 1)
		)
	).run()[0][0] or 0

	# Previous year
	previous_year = (
		frappe.qb.from_(tithes_table)
		.select(Sum(tithes_table.tithe_amount))
		.where(
			(tithes_table.date >= py_start) &
			(tithes_table.date <= py_end) &
			(tithes_table.docstatus == 1)
		)
	).run()[0][0] or 0

	month_change = calculate_percentage_change(current_month, previous_month)
	year_change = calculate_percentage_change(year_to_date, previous_year)

	return {
		"category": "Tithes",
		"current_month": current_month,
		"previous_month": previous_month,
		"year_to_date": year_to_date,
		"previous_year": previous_year,
		"month_change": month_change,
		"year_change": year_change
	}


def get_offering_data(cm_start, cm_end, pm_start, pm_end, yt_start, yt_end, py_start, py_end):
	tithes_table = frappe.qb.DocType("Tithes and Offerings")

	# Current month
	current_month = (
		frappe.qb.from_(tithes_table)
		.select(Sum(tithes_table.offering_amount))
		.where(
			(tithes_table.date >= cm_start) &
			(tithes_table.date <= cm_end) &
			(tithes_table.docstatus == 1)
		)
	).run()[0][0] or 0

	# Previous month
	previous_month = (
		frappe.qb.from_(tithes_table)
		.select(Sum(tithes_table.offering_amount))
		.where(
			(tithes_table.date >= pm_start) &
			(tithes_table.date <= pm_end) &
			(tithes_table.docstatus == 1)
		)
	).run()[0][0] or 0

	# Year to date
	year_to_date = (
		frappe.qb.from_(tithes_table)
		.select(Sum(tithes_table.offering_amount))
		.where(
			(tithes_table.date >= yt_start) &
			(tithes_table.date <= yt_end) &
			(tithes_table.docstatus == 1)
		)
	).run()[0][0] or 0

	# Previous year
	previous_year = (
		frappe.qb.from_(tithes_table)
		.select(Sum(tithes_table.offering_amount))
		.where(
			(tithes_table.date >= py_start) &
			(tithes_table.date <= py_end) &
			(tithes_table.docstatus == 1)
		)
	).run()[0][0] or 0

	month_change = calculate_percentage_change(current_month, previous_month)
	year_change = calculate_percentage_change(year_to_date, previous_year)

	return {
		"category": "Regular Offerings",
		"current_month": current_month,
		"previous_month": previous_month,
		"year_to_date": year_to_date,
		"previous_year": previous_year,
		"month_change": month_change,
		"year_change": year_change
	}


def get_special_offerings_data(cm_start, cm_end, pm_start, pm_end, yt_start, yt_end, py_start, py_end):
	tithes_table = frappe.qb.DocType("Tithes and Offerings")

	# Current month
	current_month = (
		frappe.qb.from_(tithes_table)
		.select(Sum(tithes_table.campmeeting_offering + tithes_table.church_building_offering))
		.where(
			(tithes_table.date >= cm_start) &
			(tithes_table.date <= cm_end) &
			(tithes_table.docstatus == 1)
		)
	).run()[0][0] or 0

	# Previous month
	previous_month = (
		frappe.qb.from_(tithes_table)
		.select(Sum(tithes_table.campmeeting_offering + tithes_table.church_building_offering))
		.where(
			(tithes_table.date >= pm_start) &
			(tithes_table.date <= pm_end) &
			(tithes_table.docstatus == 1)
		)
	).run()[0][0] or 0

	# Year to date
	year_to_date = (
		frappe.qb.from_(tithes_table)
		.select(Sum(tithes_table.campmeeting_offering + tithes_table.church_building_offering))
		.where(
			(tithes_table.date >= yt_start) &
			(tithes_table.date <= yt_end) &
			(tithes_table.docstatus == 1)
		)
	).run()[0][0] or 0

	# Previous year
	previous_year = (
		frappe.qb.from_(tithes_table)
		.select(Sum(tithes_table.campmeeting_offering + tithes_table.church_building_offering))
		.where(
			(tithes_table.date >= py_start) &
			(tithes_table.date <= py_end) &
			(tithes_table.docstatus == 1)
		)
	).run()[0][0] or 0

	month_change = calculate_percentage_change(current_month, previous_month)
	year_change = calculate_percentage_change(year_to_date, previous_year)

	return {
		"category": "Special Offerings",
		"current_month": current_month,
		"previous_month": previous_month,
		"year_to_date": year_to_date,
		"previous_year": previous_year,
		"month_change": month_change,
		"year_change": year_change
	}


def get_department_income_data(cm_start, cm_end, pm_start, pm_end, yt_start, yt_end, py_start, py_end):
	income_table = frappe.qb.DocType("Department Income")

	# Get departments with income
	departments = (
		frappe.qb.from_(income_table)
		.select(income_table.department)
		.distinct()
		.where(income_table.docstatus == 1)
		.orderby(income_table.department)
	).run()

	data = []
	for dept in departments:
		dept_name = dept[0]

		# Current month
		current_month = (
			frappe.qb.from_(income_table)
			.select(Sum(income_table.amount))
			.where(
				(income_table.date >= cm_start) &
				(income_table.date <= cm_end) &
				(income_table.department == dept_name) &
				(income_table.docstatus == 1)
			)
		).run()[0][0] or 0

		# Previous month
		previous_month = (
			frappe.qb.from_(income_table)
			.select(Sum(income_table.amount))
			.where(
				(income_table.date >= pm_start) &
				(income_table.date <= pm_end) &
				(income_table.department == dept_name) &
				(income_table.docstatus == 1)
			)
		).run()[0][0] or 0

		# Year to date
		year_to_date = (
			frappe.qb.from_(income_table)
			.select(Sum(income_table.amount))
			.where(
				(income_table.date >= yt_start) &
				(income_table.date <= yt_end) &
				(income_table.department == dept_name) &
				(income_table.docstatus == 1)
			)
		).run()[0][0] or 0

		# Previous year
		previous_year = (
			frappe.qb.from_(income_table)
			.select(Sum(income_table.amount))
			.where(
				(income_table.date >= py_start) &
				(income_table.date <= py_end) &
				(income_table.department == dept_name) &
				(income_table.docstatus == 1)
			)
		).run()[0][0] or 0

		month_change = calculate_percentage_change(current_month, previous_month)
		year_change = calculate_percentage_change(year_to_date, previous_year)

		data.append({
			"category": f"{dept_name} Income",
			"current_month": current_month,
			"previous_month": previous_month,
			"year_to_date": year_to_date,
			"previous_year": previous_year,
			"month_change": month_change,
			"year_change": year_change
		})

	return data


def get_expense_data(cm_start, cm_end, pm_start, pm_end, yt_start, yt_end, py_start, py_end):
	expense_table = frappe.qb.DocType("Department Expense")

	# Get expenses by department
	departments = (
		frappe.qb.from_(expense_table)
		.select(expense_table.department)
		.distinct()
		.where(expense_table.docstatus == 1)
		.orderby(expense_table.department)
	).run()

	data = []
	for dept in departments:
		dept_name = dept[0]

		# Current month
		current_month = (
			frappe.qb.from_(expense_table)
			.select(Sum(expense_table.total_amount))
			.where(
				(expense_table.expense_date >= cm_start) &
				(expense_table.expense_date <= cm_end) &
				(expense_table.department == dept_name) &
				(expense_table.docstatus == 1)
			)
		).run()[0][0] or 0

		# Previous month
		previous_month = (
			frappe.qb.from_(expense_table)
			.select(Sum(expense_table.total_amount))
			.where(
				(expense_table.expense_date >= pm_start) &
				(expense_table.expense_date <= pm_end) &
				(expense_table.department == dept_name) &
				(expense_table.docstatus == 1)
			)
		).run()[0][0] or 0

		# Year to date
		year_to_date = (
			frappe.qb.from_(expense_table)
			.select(Sum(expense_table.total_amount))
			.where(
				(expense_table.expense_date >= yt_start) &
				(expense_table.expense_date <= yt_end) &
				(expense_table.department == dept_name) &
				(expense_table.docstatus == 1)
			)
		).run()[0][0] or 0

		# Previous year
		previous_year = (
			frappe.qb.from_(expense_table)
			.select(Sum(expense_table.total_amount))
			.where(
				(expense_table.expense_date >= py_start) &
				(expense_table.expense_date <= py_end) &
				(expense_table.department == dept_name) &
				(expense_table.docstatus == 1)
			)
		).run()[0][0] or 0

		month_change = calculate_percentage_change(current_month, previous_month)
		year_change = calculate_percentage_change(year_to_date, previous_year)

		data.append({
			"category": f"{dept_name} Expenses",
			"current_month": current_month,
			"previous_month": previous_month,
			"year_to_date": year_to_date,
			"previous_year": previous_year,
			"month_change": month_change,
			"year_change": year_change
		})

	return data


def calculate_total_income(data):
	# Calculate totals from income items (skip headers)
	income_items = [item for item in data if item.get("category") and
					not item["category"].startswith("<b>") and
					(item["category"] in ["Tithes", "Regular Offerings", "Special Offerings"] or
					 item["category"].endswith(" Income"))]

	current_month = sum(flt(item.get("current_month", 0)) for item in income_items)
	previous_month = sum(flt(item.get("previous_month", 0)) for item in income_items)
	year_to_date = sum(flt(item.get("year_to_date", 0)) for item in income_items)
	previous_year = sum(flt(item.get("previous_year", 0)) for item in income_items)

	month_change = calculate_percentage_change(current_month, previous_month)
	year_change = calculate_percentage_change(year_to_date, previous_year)

	return {
		"category": "<b>TOTAL INCOME</b>",
		"current_month": current_month,
		"previous_month": previous_month,
		"year_to_date": year_to_date,
		"previous_year": previous_year,
		"month_change": month_change,
		"year_change": year_change
	}


def calculate_percentage_change(current, previous):
	if not previous:
		return 0
	return ((current - previous) / previous) * 100


def get_chart_data(data, filters):
	"""Generate chart data for financial summary"""
	if not data:
		return None

	# Extract income and expense data
	income_data = []
	expense_data = []

	for row in data:
		category = row.get("category", "")
		if not category or category.startswith("<b>") or not category.strip():
			continue

		current_month = row.get("current_month", 0)
		year_to_date = row.get("year_to_date", 0)

		if category in ["Tithes", "Regular Offerings", "Special Offerings"]:
			income_data.append({
				"category": category,
				"current_month": current_month,
				"year_to_date": year_to_date
			})
		elif "Expenses" in category:
			expense_data.append({
				"category": category,
				"current_month": current_month,
				"year_to_date": year_to_date
			})

	# Create chart data for income vs expenses comparison
	income_total = sum(item["current_month"] for item in income_data)
	expense_total = sum(item["current_month"] for item in expense_data)

	return {
		"data": {
			"labels": ["Income", "Expenses"],
			"datasets": [{
				"values": [income_total, expense_total]
			}]
		},
		"type": "pie",
		"height": 300,
		"colors": ["#4CAF50", "#F44336"]
	}


@frappe.whitelist()
def get_income_breakdown_chart(data, filters):
	"""Generate income breakdown chart data"""
	if not data:
		return None

	# Parse JSON data if it's a string
	import json
	if isinstance(data, str):
		try:
			data = json.loads(data)
		except:
			return None

	if not isinstance(data, list):
		return None

	income_categories = []
	income_values = []

	for row in data:
		if not isinstance(row, dict):
			continue

		category = row.get("category", "")
		if category in ["Tithes", "Regular Offerings", "Special Offerings"]:
			income_categories.append(category)
			# Convert to float, handling empty strings and None values
			try:
				value = float(row.get("current_month", 0) or 0)
			except (ValueError, TypeError):
				value = 0
			income_values.append(value)

	return {
		"data": {
			"labels": income_categories,
			"datasets": [{
				"values": income_values
			}]
		},
		"type": "donut",
		"height": 300,
		"colors": ["#2E7D32", "#1976D2", "#F57C00"]
	}


@frappe.whitelist()
def get_expense_breakdown_chart(data, filters):
	"""Generate expense breakdown chart data"""
	if not data:
		return None

	# Parse JSON data if it's a string
	import json
	if isinstance(data, str):
		try:
			data = json.loads(data)
		except:
			return None

	if not isinstance(data, list):
		return None

	expense_categories = []
	expense_values = []

	for row in data:
		if not isinstance(row, dict):
			continue

		category = row.get("category", "")
		if "Expenses" in category:
			expense_categories.append(category.replace(" Expenses", ""))
			# Convert to float, handling empty strings and None values
			try:
				value = float(row.get("current_month", 0) or 0)
			except (ValueError, TypeError):
				value = 0
			expense_values.append(value)

	return {
		"data": {
			"labels": expense_categories,
			"datasets": [{
				"values": expense_values
			}]
		},
		"type": "bar",
		"height": 300,
		"colors": ["#D32F2F", "#7B1FA2", "#F57C00", "#388E3C"]
	}


@frappe.whitelist()
def get_trend_comparison_chart(data, filters):
	"""Generate trend comparison chart data"""
	if not data:
		return None

	# Parse JSON data if it's a string
	import json
	if isinstance(data, str):
		try:
			data = json.loads(data)
		except:
			return None

	if not isinstance(data, list):
		return None

	categories = []
	current_month_values = []
	previous_month_values = []
	ytd_values = []

	for row in data:
		if not isinstance(row, dict):
			continue

		category = row.get("category", "")
		if category in ["Tithes", "Regular Offerings", "Special Offerings"]:
			categories.append(category)
			# Convert to float, handling empty strings and None values
			try:
				current_value = float(row.get("current_month", 0) or 0)
			except (ValueError, TypeError):
				current_value = 0
			current_month_values.append(current_value)

			# Convert previous month value
			try:
				previous_value = float(row.get("previous_month", 0) or 0)
			except (ValueError, TypeError):
				previous_value = 0
			previous_month_values.append(previous_value)

			# Convert YTD value
			try:
				ytd_value = float(row.get("year_to_date", 0) or 0)
			except (ValueError, TypeError):
				ytd_value = 0
			ytd_values.append(ytd_value)

	return {
		"data": {
			"labels": categories,
			"datasets": [
				{
					"name": "Current Month",
					"values": current_month_values
				},
				{
					"name": "Previous Month",
					"values": previous_month_values
				},
				{
					"name": "Year to Date",
					"values": ytd_values
				}
			]
		},
		"type": "bar",
		"height": 350,
		"colors": ["#4CAF50", "#FF9800", "#2196F3"]
	}
